src/server/acl/acl_family.cc

```diff
@@ -57,10 +61,113 @@ void AclFamily::List(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+namespace {
+
+std::optional<std::string> MaybeParsePassword(std::string_view command) {
+  if (command[0] != '>') {
+    return {};
+  }
+
+  return {std::string(command.substr(1))};
+}
+
+std::optional<bool> MaybeParseStatus(std::string_view command) {
+  if (command == "ON") {
+    return true;
+  }
+  if (command == "OFF") {
+    return false;
+  }
+  return {};
+}
+
+using OptCat = std::optional<uint32_t>;
+
+std::pair<OptCat, OptCat> MaybeParseAclCategory(std::string_view command) {
+  if (command[0] == '+' && command[1] == '@') {
+    auto res = CATEGORY_INDEX_TABLE.find(command.substr(2));
+    if (res == CATEGORY_INDEX_TABLE.end()) {
+      return {};
+    }
+    return {res->second, {}};
+  }
+
+  if (command[0] == '-' && command[1] == '@') {
+    auto res = CATEGORY_INDEX_TABLE.find(command.substr(2));
+    if (res == CATEGORY_INDEX_TABLE.end()) {
+      return {};
+    }
+    return {{}, res->second};
+  }
+
+  return {};
+}
+
+std::variant<User::UpdateRequest, std::string> ParseAclSetUser(CmdArgList args) {
+  User::UpdateRequest req;
+
+  auto maybe_update_acl = [](auto& output, const auto& input) {
+    if (input) {
+      if (!output) {
+        output = *input;
+      }
+      *output |= *input;
+    }
+  };
+
+  for (auto arg : args) {
+    ToUpper(&arg);
+    const auto command = facade::ToSV(arg);
+    if (auto pass = MaybeParsePassword(command); pass) {
+      if (req.password) {
+        return "Only one password is allowed";
+      }
+      req.password = std::move(pass);
+      continue;
+    }
+
+    if (auto status = MaybeParseStatus(command); status) {
+      if (req.is_active) {
+        return "Multiple ON/OFF are not allowed";
+      }
+      req.is_active = *status;
+      continue;
+    }
+
+    auto [acl_plus, acl_minus] = MaybeParseAclCategory(command);
+    if (!acl_plus && !acl_minus) {
+      using namespace std::string_literals;
+      return "Unrecognized parameter: "s + std::string(facade::ToSV(arg));
+    }
+
+    maybe_update_acl(req.plus_acl_categories, acl_plus);
+    maybe_update_acl(req.minus_acl_categories, acl_minus);
+  }
+
+  return req;
+}
+
+}  // namespace
+
+void AclFamily::SetUser(CmdArgList args, ConnectionContext* cntx) {
+  std::string_view username = facade::ToSV(args[0]);
+  auto req = ParseAclSetUser(args.subspan(1));
+  auto error_case = [cntx](std::string&& error) { (*cntx)->SendError(error); };
+  auto update_case = [username, cntx](User::UpdateRequest&& req) {
+    ServerState::tlocal()->user_registry->MaybeAddAndUpdate(username, std::move(req));
+    (*cntx)->SendOk();
+  };
+
+  std::visit(Overloaded(error_case, update_case), std::move(req));
```

Comment:
Unusual to see this but I like it more than that we generally use

We generally do `optional<ResultType> ParseXXXOrReply()` so we don't have to pass around the error. I don't see it being a problem because those functions are very unlikely to be ever re-used.

Small note: Since recently we have the ErrorReply type that you can use instead of a string